src/providers/Auth.tsx

```diff
@@ -0,0 +1,127 @@
+import Cookies from 'js-cookie';
+import { EGO_JWT_KEY, RDPC_DOMAIN, RDPC_REGION } from 'config/globals';
+import { decodeToken, isValidJwt, getPermissionsFromToken, isDccMember as isDccMemberUtil } from 'utils/egoJwt';
+
+type T_EgoToken = string;
+
+type T_AuthContext = [
+  [T_EgoToken, React.Dispatch<React.SetStateAction<T_EgoToken>>],
+];
+
+export const AuthContext = React.createContext<T_AuthContext>([['', () => {}]]);
+
+export const AuthProvider = ({ children }: any) => {
+  const [token, setToken] = useState<T_EgoToken>('');
+
+  return (
+    <AuthContext.Provider value={[[token, setToken]]}>
+      {children}
+    </AuthContext.Provider>
+  );
+};
+
+export const useAuth = () => {
+  const [
+    [token, setTokenState],
+  ] = useContext(AuthContext);
+
+  const getPermissions = (token?: string) => {
+    if (token) {
+      return getPermissionsFromToken(token) || [];
+    }
+
+    return getPermissionsFromToken(getToken()) || [];
+  };
+
+  const canRead = (token?: string) => {
+    if (token) {
+      return getPermissionsFromToken(token).filter(permission => permission.toLowerCase().startsWith(`${RDPC_DOMAIN}-${RDPC_REGION}.READ`.toLowerCase())).length > 0;
+    }
+
+    return getPermissions().filter(permission => permission.toLowerCase().startsWith(`${RDPC_DOMAIN}-${RDPC_REGION}.READ`.toLowerCase())).length > 0;
+  };
+
+  const canWrite = (token?: string) => {
+    if (token) {
+      return getPermissionsFromToken(token).filter(permission => permission.toLowerCase().startsWith(`${RDPC_DOMAIN}-${RDPC_REGION}.WRITE`.toLowerCase())).length > 0;
+    }
+
+    return getPermissions().filter(permission => permission.toLowerCase().startsWith(`${RDPC_DOMAIN}-${RDPC_REGION}.WRITE`.toLowerCase())).length > 0;
+  };
+
+  const isDccMember = (token?: string) => {
+    if (token) {
+      return isDccMemberUtil(getPermissions(token));
+    }
+
+    return isDccMemberUtil(getPermissions());
+  };
+
+  const isLoggedIn = (): boolean => {
+    return isValidJwt(getToken());
+  };
+
+  const getToken = (): string => {
+    if (!token && !Cookies.get(EGO_JWT_KEY)) {
+      return '';
+    }
+    
+    if (!token && Cookies.get(EGO_JWT_KEY)) {
+      return Cookies.get(EGO_JWT_KEY) || '';
+    }
+
+    return token;
+  };
+
+  const setToken = (token: string) => {
+    Cookies.set(EGO_JWT_KEY, token);
+    setTokenState(token);
+  };
+
+  const clearToken = () => {
+    Cookies.remove(EGO_JWT_KEY);
+    setTokenState('');
+  };
+
+  const getUserModel = () => {
+    if (isValidJwt(getToken())) {
+      const data = decodeToken(getToken());
+      if (data && data.context && data.context.user) {
+        return data.context.user;
+      }
+    }
+
+    return null;
+  };
+
+  return {
+    token: getToken(),
+    setToken,
+    clearToken,
+    permissions: getPermissions(),
+    userModel: getUserModel(),
+    isLoggedIn: isLoggedIn(),
+    isMember: canRead,
+    isAdmin: canWrite,
+    isDccMember,
```

Comment:
why expose these as functions rather than just compute it once and expose the result? since everything is derived from `token` which is a react state, it should auto update when `token` changes eh.

Unless there's a use case for authorizing more than one token at a time in the app, but that seems unlikely.

Reply:
Originally I had it doing just that, but I ran into some issues where I wanted to check the status of isAdmin/isMember/isDccMember at the same time as or immediately after setting the token, but because of the react lifecyle the values weren't updated until the next tick so they'd still be false.

Namely, in `LoggedIn.tsx`, when checking the permissions of the user upon login before redirecting them to the appropriate route. I ended up exposing these functions so I could pass in the JWT and check its validity and whether the user even has access to the site before storing it. This way, only if the user has access to the site do we even bother storing their token (get get kicked to the "No Access" page if they are not an RDPC member, admin, or DCC member)